### PR TITLE
feat: support external udaf function

### LINKED ISSUE #1835: Support creating aggregate function
we can only add udf function with `CREATE FUNCTION` now.

### PR DESCRIPTION
**Pull request by dl239**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

* **What is the current behavior?** (You can also link to an open issue here)

* **What is the new behavior (if this is a feature change)?**

### CODE REVIEW

**Comment by dl239** [comment]
support null value

File: docs/en/developer/udf_develop_guide.md [hunk 1]
```diff
@@ -3,7 +3,7 @@
 Although there are already hundreds of built-in functions, they can not satisfy the needs in some cases. In the past, this could only be done by developing new built-in functions. Built-in function development requires a relatively long cycle because it needs to recompile binary files and users have to wait for new version release.
 In order to help users to quickly develop computing functions that are not provided by OpenMLDB, we develop the mechanism of user dynamic registration function.
 
-SQL functions can be categorised into single-line functions and aggregate functions. At present, only single-line UDFs are supported. An introduction to single-line functions and aggregate functions can be seen [here](./built_in_function_develop_guide.md).
+SQL functions can be categorised into single-line functions and aggregate functions. An introduction to single-line functions and aggregate functions can be seen [here](./built_in_function_develop_guide.md).
```
Comment (aceforeverd) [inline docs/en/developer/udf_develop_guide.md]: single-line function is strange. Maybe `scalar function`

File: docs/en/developer/udf_develop_guide.md
Comment (aceforeverd) [inline docs/en/developer/udf_develop_guide.md:1]: Need a summary about how user defined function works. 

For builtin functions, internally, it supports register function via:
1. external function
2. Expr Node
3. codegen logic
4. and maybe above three methods with cpp templates

I see currently, for user defined function, it only work with external functions (without templates)

File: hybridse/include/node/node_enum.h [hunk 2]
```diff
@@ -93,7 +93,6 @@ enum SqlNodeType {
     kDeleteStmt,
     kCreateFunctionStmt,
     kDynamicUdfFnDef,
-    kDynamicUdafFnDef,
```
Comment (aceforeverd) [inline hybridse/include/node/node_enum.h:96]: rm will break `kWithClauseEntry` with old version

File: docs/en/developer/udf_develop_guide.md [hunk 3]
```diff
@@ -44,17 +45,28 @@ The SQL types corresponding to C++ types are shown as follows:
 ```c++
     struct UDFContext {
         ByteMemoryPool* pool;  // Used for memory allocation.
-        void* ptr;             // Used for the storage of temporary variables. Single-line functions don't use it at present.
+        void* ptr;             // Used for the storage of temporary variables for aggregrate functions.
```
Comment (aceforeverd) [inline docs/en/developer/udf_develop_guide.md:48]: `void*` is not type safe ...

File: docs/zh/tutorial/udf_develop_guide.md [hunk 4]
```diff
@@ -44,13 +45,25 @@
 
 #### 2.1.4 内存管理
 
-- 在自定义函数中，不允许使用`new`操作符或者`malloc`函数开辟空间。
-- 若需要动态开辟空间，需要使用OpenMLDB提供的内存管理接口。函数执行完OpenMLDB会自动释放内存.
+- 在单行函数中，不允许使用`new`操作符或者`malloc`函数开辟空间，当然也不需要用`free`或者`delete`释放。
+- 在聚合函数中，在init函数中可以使用`new`/`malloc`开辟空间，但是必须在output函数中释放。最后的返回值如果是string需要保存在mempool开辟的空间中
```
Comment (aceforeverd) [inline docs/zh/tutorial/udf_develop_guide.md:49]: 涉及到 params & return value 的地方不可以用 malloc 操作，但是计算过程中的中间变量因为生命周期不会超过 SQL function 的范围所以用 malloc 也没限制。

例如 C type of `void func(T1* param1, T2* param2, T3* output)` 这样，要将计算结果写入 `output`, 那就不建议 malloc 开辟空间，因为 `output` 还会在外部访问，到时候就不知道如何 `free`. 但如果你只是在 这个 function definition 里面定义了临时变量，malloc 即可。

File: docs/en/reference/sql/ddl/CREATE_FUNCTION.md [hunk 5]
```diff
@@ -43,4 +43,14 @@ CREATE FUNCTION cut2(x string) RETURNS string OPTIONS (FILE = 'libtest_udf.so');
...
+```
+
+Create an aggregate function whose input parameter type is bigint, the return type is bigint, and the dynamic library file is libtest_udf.so
+```sql
+CREATE AGGREGATE FUNCTION special_sum(x BIGINT) RETURNS BIGINT OPTIONS (FILE = 'libtest_udf.so');
+```
+
+Create an aggregate function whose input parameter type is bigint and nullable, the return type is bigint and nullable, and the dynamic library file is libtest_udf.so
+```sql
+CREATE AGGREGATE FUNCTION count_null(x BIGINT) RETURNS BIGINT OPTIONS (FILE = 'libtest_udf.so', ARG_NULLABLE=true, RETURN_NULLABLE=true);
```
Comment (aceforeverd) [inline docs/en/reference/sql/ddl/CREATE_FUNCTION.md:55]: Maybe we can deduce the nullable info from the external function define, so the two nullable flags can omited.

File: hybridse/include/node/node_enum.h (hunk 2 above)
Comment (dl239) [reply hybridse/include/node/node_enum.h:96]: what's the impact

File: docs/en/reference/sql/ddl/CREATE_FUNCTION.md (hunk 5 above)
Comment (dl239) [reply docs/en/reference/sql/ddl/CREATE_FUNCTION.md:55]: there is the way to get the function prototype from dynamic lib ?

File: docs/en/developer/udf_develop_guide.md (hunk 3 above)
Comment (dl239) [reply docs/en/developer/udf_develop_guide.md:48]: please left you advice

File: docs/zh/tutorial/udf_develop_guide.md (hunk 4 above)
Comment (dl239) [reply docs/zh/tutorial/udf_develop_guide.md:49]: 是的

File: docs/en/developer/udf_develop_guide.md
Comment (dl239) [reply docs/en/developer/udf_develop_guide.md:1]: this is a usage documents, we can write a technical docs to introduce more details

File: hybridse/include/node/node_enum.h (hunk 2 above)
Comment (aceforeverd) [reply hybridse/include/node/node_enum.h:96]: value of `kWithClauseEntry` changes

File: docs/en/developer/udf_develop_guide.md
Comment (aceforeverd) [reply docs/en/developer/udf_develop_guide.md:1]: Just a simple line to express the register progress is sufficient.  like 
> openmldb dynamic loading the compiled library contains user defined function

?

File: docs/en/developer/udf_develop_guide.md (hunk 3 above)
Comment (aceforeverd) [reply docs/en/developer/udf_develop_guide.md:48]: No idea... But user has more rate to write a unsafe code and boom themself. :(

File: docs/en/reference/sql/ddl/CREATE_FUNCTION.md (hunk 5 above)
Comment (aceforeverd) [reply docs/en/reference/sql/ddl/CREATE_FUNCTION.md:55]: possible in cpp with template deduction (guesing). Just further enhancements

File: include/udf/openmldb_udf.h
Comment (aceforeverd) [inline include/udf/openmldb_udf.h:1]: `using` in a header file is not advised, better remove it

File: hybridse/include/node/node_enum.h (hunk 2 above)
Comment (dl239) [reply hybridse/include/node/node_enum.h:96]: i think there is no effect of the behavior of the program although the values change. if have, please let me known

File: docs/en/developer/udf_develop_guide.md
Comment (dl239) [reply docs/en/developer/udf_develop_guide.md:1]: ok

**Review by lqy222 — APPROVED** [review]

**Review by aceforeverd — APPROVED** [review]

**Pull request merged**